jena-db/jena-dboe-storage/src/main/java/org/apache/jena/dboe/storage/system/DatasetGraphTxnCtl.java

```diff
@@ -0,0 +1,353 @@
+    }
+
+    // Any writers - WRITE, READ_PROMOTE, READ_COMMITTED_PROMOTE
+    // Writer take this lock (slightly confusingly as a "read" lock)
+    // READ-ONLY operation do no tneed to take this block.
+
+    protected void startReadOnly() {
+        startReadOnly(true);
+    }
+
+    /**
+     * Indicate the database is to be readonly.
+     * Any writer, or potential writer, takes this lock in "multi" mode.
+     * To block, call this method.
+     */
+    protected void startReadOnly(boolean canBlock) {
+        beginSingleMode(writeableDatabase, canBlock);
+    }
+
+    protected void finishReadOnly() {
+        endSingleMode(writeableDatabase);
+    }
+
+    protected void startPossibleWriter() {
+        beginMultiMode(writeableDatabase, true);
+    }
+
+    protected void startPossibleWriter(boolean canBlock) {
+        beginMultiMode(writeableDatabase, canBlock);
+    }
+
+    protected void finishPossibleWriter() {
+        endMultiMode(writeableDatabase);
+    }
+
+    // Lock abstraction.
+    /* Function for enter-leave for two lock types:
+     *    Multiple active threads or exclusive access. (MRSW)
+     *      beginMultiMode, beginSingleMode
+     *    Multiple active threads and one distinguished thread (MR+SW).
+     *      beginSingleGate
+     *
+     * == Multiple active threads or exclusive access. (MRSW)
+     * This uses a ReadWriteLock where "read" is multiple outstanding threads
+     * and "write" is exclusive access.
+     * Used for:
+     *   Exclusivity lock
+     *   No potential writer lock (TxnType.WRITE or PROMOTE).
+     *
+     * The use of R and W for the exclusive access locks are slightly confusing.
+     * It is not related to whether a transaction is R or W.
+     * R means "multiple", W means "exclusive"
+     *
+     * == Multiple active threads and one distinguished thread (MR+SW).
+     * This is a semaphore with one permit taken by the distinguished thread.
+     * The other active threads are not affected.
+     * Use for:
+     *   Multiple ReadWrite==READ and single ReadWrite.WRITE
+     */
+
+    // MRSW - two modes: multiple OR a single thread.
+
+
+
+    // -- Exclusivity
+
+    @Override
+    public void startNonExclusiveMode() {
+        boolean b = beginMultiMode(exclusivitylock, true);
+        if ( !b )
+            throw new TransactionException("Can't start transaction at the moment.");
+    }
+
+    @Override
+    public boolean tryNonExclusiveMode(boolean canBlock) {
+        return beginMultiMode(exclusivitylock, canBlock);
+    }
+
+    @Override
+    public void finishNonExclusiveMode() {
+        endMultiMode(exclusivitylock);
+    }
+
+    /**
+     * Enter exclusive mode; block if necessary. There are no active transactions on
+     * return; new transactions will be held up in 'begin'. Return to normal (release
+     * waiting transactions, allow new transactions) with
+     * {@link #finishExclusiveMode}.
+     * <p>
+     * Do not call inside an existing transaction.
```

Comment:
Isn't there a way to the code to tell whether it's running inside a transaction at this point? I think, if I recall correctly, I've seen some code doing that in.. JPA? Some Spring transaction code? But I am not sure how deep this code is in the execution context, nor if it'd be possible to expose the transaction at this point. Just curious :grimacing:

Reply:
In doing this bug fix, I scaled back from doing a large rework of transaction control. Not a good time just before a release!

There is `isInTransaction()` but at this point the real transaction which is on the storage database isn't started. I'm not sure the sequencing works. `DatasetGraphCxnCtl` is only controlling entry and exit to the switchable. The real transaction is in the storage `DatasetGraphTDB`. I'd like to split `TransactionCcordinator` into parts, the controls and the transaction mechanism itself.  I copied a more general `DatasetGraphCxnCtl` to a scratch area for later; it was getting too big a change just for the bug.

This is very low level messing around with the internals of the transaction system. An exception isn't great as it will naturally propagate out without other clearup. Maybe the javadoc should say "Don't call this operation" :smile:.

`isInTransaction` uses a ThreadLocal. ThreadLocals are easy to get wrong and slowly leak memory in a long running process (you have to explicitly remove them if threads are being created and which finish).

Reply:
>In doing this bug fix, I scaled back from doing a large rework of transaction control. Not a good time just before a release!
> (…) it was getting too big a change just for the bug.

:+1: sounds reasonable!

>This is very low level messing around with the internals of the transaction system. An exception isn't great as it will naturally propagate out without other clearup. Maybe the javadoc should say "Don't call this operation" smile.

I don't believe we will get many users navigating this deep into the code. We can leave the comment as-is now. Looks like you have a plan for improvements that may alter parts of this code too—and will probably change the comments anyway.

>isInTransaction uses a ThreadLocal. ThreadLocals are easy to get wrong and slowly leak memory in a long running process (you have to explicitly remove them if threads are being created and which finish).

:+1: good to be reminded of that! Every time I switch from JS or Python to Java I find that I forget things about parts that Java do not have in common with these languages (like threads, jee, memory management, etc) :bow:

Reply:
It's a nuisance these operations have to be public (short of JPMS modules). Really, it would be better with "friends of" declarations to span boundaries without being completely public. But the transaction system is not specific to TDB2 and so it's in a separate package and maven module.